### PR TITLE
Create sample loader out of user's data loader

### LINKED ISSUE #779: [FEAT][CV] Sample from dataloader and cache label and predictions
**Is your feature request related to a problem? Please describe.**
Currently vision checks may take a very long time to iterate over all the data.

**Describe the solution you'd like**

1. We should be able to randomly sample from a dataloader to get a fixed amount of samples.
2. We should cache the labels and predictions for these samples in the context object (but of course should be able to collect them anew when running a specific check not in a suite).

**Describe alternatives you've considered**
Caching the labels and predictions is optional

### PR DESCRIPTION
**Pull request by matanper**
We want to have a data loader with only a sample of the data, and have it return a constant order each time to allow us to cache predictions.

resolve #779

### CODE REVIEW

File: deepchecks/vision/dataset.py [hunk 1]
```diff
@@ -179,3 +188,81 @@ def validate_dataset(cls, obj) -> 'VisionDataset':
             raise DeepchecksValueError('Check requires a non-empty dataset')
 
         return obj
+
```
Comment (nirhutnik) [inline deepchecks/vision/dataset.py:224]: Add a public function to call the _sample_data generator.

File: deepchecks/vision/dataset.py [hunk 2]
```diff
@@ -179,3 +188,81 @@ def validate_dataset(cls, obj) -> 'VisionDataset':
...
+            return DataLoader(samples_dataset, generator=generator(), sampler=SequentialSampler(samples_data),
+                              **common_props_to_copy)
+        else:
+            length = len(dataset)
+            return DataLoader(dataset, generator=generator(),
+                              sampler=FixedSampler(length, seed, sample_size), **common_props_to_copy)
+
+
+class InMemoryDataset(Dataset):
+    """Dataset implementation that gets all the data as in-memory list."""
```
[the remark this answers is not in the document]
Comment (matanper) [reply deepchecks/vision/dataset.py:244]: Exception I guess

File: deepchecks/vision/dataset.py (hunk 1 above)
Comment (matanper) [reply deepchecks/vision/dataset.py:241]: why? where else it will be used?

Comment (nirhutnik) [reply deepchecks/vision/dataset.py:241]: Left this before I saw you added caching. Thought this will be called from outside

File: deepchecks/vision/dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/vision/dataset.py:244]: Don't we want our own exception in this case? Saying that the sample given is too large?

[the remark this answers is not in the document]
Comment (matanper) [reply deepchecks/vision/dataset.py:244]: how do we measure a too large sample?

**Comment by ItayGabbay** [comment]
This PR contains many unrelated changes, I am not sure why. Could you please create a new one from the main branch?

**Comment by matanper** [comment]
PR got dirty, opened a new cleaner one
